packages/react-components/components/SmartTopAlert.tsx

```diff
@@ -1,101 +1,181 @@
-  cachedText: string | null
-  visible: boolean
-}
-
 interface Props {
   title?: string | null
   text: string | null
   onPress: () => void
   type: NotificationTypes
-  dismissAfter: number | null
+  dismissAfter?: number | null
   buttonMessage?: string | null
 }
 
-class SmartTopAlert extends React.Component<Props, State> {
-  timeout: number | null = null
+// This component needs to be always mounted for the hide animation to be visible
+function SmartTopAlert(props: Props) {
+  const [visibleAlertState, setVisibleAlertState] = useState<Props | null>(null)
+  const yOffset = useRef(new Animated.Value(-500))
+  const containerRef = useRef<View>()
+  const animatedRef = useCallback((node) => {
+    containerRef.current = node && node.getNode()
+  }, [])
+
+  const alertState = useMemo(
+    () => {
+      // tslint bug?
+      // tslint:disable-next-line: no-shadowed-variable
+      const { type, title, text, buttonMessage, dismissAfter, onPress } = props
+      if (title || text) {
+        return {
+          type,
+          title,
+          text,
+          buttonMessage,
+          dismissAfter,
+          onPress,
+        }
+      } else {
+        return null
+      }
+    },
+    [props.type, props.title, props.text, props.buttonMessage, props.dismissAfter, props.onPress]
+  )
 
-  constructor(props: Props) {
-    super(props)
-    this.state = {
-      cachedTitle: props.title,
-      cachedText: props.text,
-      visible: !!props.text,
+  function hide() {
+    if (containerRef.current) {
+      containerRef.current.measure((l, t, w, height) => {
+        Animated.timing(yOffset.current, {
+          toValue: -height,
+          duration: 150,
+          useNativeDriver: true,
+        }).start(({ finished }) => {
+          if (finished) {
+            setVisibleAlertState(null)
+          }
+        })
+      })
     }
   }
 
-  componentDidUpdate(prevProps: Props) {
-    if (this.props.text && prevProps.text !== this.props.text) {
-      this.setState({ cachedText: this.props.text, cachedTitle: this.props.title })
-      if (this.timeout) {
-        clearTimeout(this.timeout)
-        this.timeout = null
+  useEffect(
+    () => {
+      if (alertState) {
+        // show
+        setVisibleAlertState(alertState)
+      } else {
+        // hide
+        hide()
+      }
+    },
+    [alertState]
+  )
+
+  useEffect(
+    () => {
+      let rafHandle: number
+      let timeoutHandle: number
+
+      if (visibleAlertState) {
+        rafHandle = requestAnimationFrame(() => {
+          if (containerRef.current) {
+            containerRef.current.measure((l, t, w, height) => {
+              Animated.timing(yOffset.current, {
+                // @ts-ignore, react-native type defs are missing this one!
+                fromValue: -height,
+                toValue: 0,
+                duration: 200,
+                useNativeDriver: true,
+              }).start()
+
+              if (visibleAlertState.dismissAfter) {
```

Comment:
This may just be a limitation of a hooks based component but generally when I see 3+ nested `if` blocks, that's a sign of a problem. Is there a better way to organize this?

Reply:
I agree, this would be nicer if `requestAnimationFrame` and `measure` would be promises.
I could wrap those and turn them into promises in a utils/helper file.
What do you think?

Reply:
Having second thoughts, I could also just return early so it looks less nested.

Reply:
I went with the return early solution for now.